AI #2359: Add Actor (`Consumer ID` and `Principal ID`) columns to `Cost and Usage` dataset (#2360)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/consumerid.md

# Consumer ID

A Consumer ID is an identifier representing the [*consumer*](#glossary:consumer): a discrete downstream entity (e.g., end-user, tenant, application, autonomous agent, human actor) that consumed a [*resource*](#glossary:resource) or [*service*](#glossary:service). The Consumer ID is commonly used for user-level cost allocation, showback, and unit economics calculations without exposing raw actor data in the central billing dataset.

> Consumer ID works in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Principal ID captures the authenticated system or service role that initiated a request at the infrastructure layer, Consumer ID captures the logical downstream actor at the application layer.

Author: ```suggestion
Consumer ID works in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Principal ID captures the authenticated system or service actor that initiated a request at the infrastructure layer, Consumer ID captures the logical downstream actor at the application layer.
```

> Consumer ID works in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Principal ID captures the authenticated actor that initiated a request at the infrastructure layer, Consumer ID captures the logical downstream actor at the application layer. For scenarios demonstrating how these columns are populated across different technology categories, see [Examples: Actor Attribution](#appendix.examples:actorattribution).

Reviewer: ```suggestion
A Consumer ID is an identifier representing the [*consumer*](#glossary:consumer): a discrete downstream actor (e.g., end-user, application, autonomous agent) that consumed a [*resource*](#glossary:resource) or [*service*](#glossary:service). The Consumer ID is commonly used for cost allocation, showback/chargeback, and unit economics calculations without exposing raw actor data.

Consumer ID is used in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Consumer ID represents the downstream actor that consumed the resource or service, Principal ID represents the authenticated and authorized actor that initiated a request.

For scenarios demonstrating how these columns are populated across different technology categories, see [Examples: Actor Attribution](#appendix.examples:actorattribution).
```

> > Note: While Consumer ID is designed to capture opaque identifiers rather than plain-text names or email addresses, these values may still be classified as Personal Data or Personally Identifiable Information (PII) under privacy frameworks such as GDPR or CCPA (e.g., as pseudonymized data). Organizations need to separately ensure that the ingestion, storage, and processing of datasets containing this column comply with their internal data privacy, security, and retention policies.

Reviewer 2: ```suggestion
> Important Consideration: While Consumer ID is designed to capture opaque identifiers rather than plain-text names or email addresses, these values may still be classified as Personal Data or Personally Identifiable Information (PII) under privacy frameworks such as GDPR or CCPA (e.g., as pseudonymized data). Organizations need to separately ensure that the ingestion, storage, and processing of datasets containing this column comply with their internal data privacy, security, and retention policies.
```

Per [AGENTS.md](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/AGENTS.md) ("Important notes must use the blockquote format `> Important Consideration`") and existing prior art in [commitment_program_eligibility_details.md:44](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/supported_features/commitment_program_eligibility_details.md#L44). The `> Note:` pattern is reserved for the section-level "informative and non-normative" disclaimer (see overview.md, appendix_overview.md).

Author: This is incorrect.  We just merged AGENTS.md into `working_draft` yesterday that changed this to be `Note` instead of `Important Consideration`.  Tricky part is that `working_draft_interim` and `working_draft` are now out of sync.  Let me touch base with @jpradocueva on this.

Author: After discussion: we will keep `working_draft_interim` and `working_draft` in sync starting this coming Thu, once we have merged something valid into `working_draft_interim`.  Closing with no action.

## Requirements

ConsumerId MUST adhere to the following requirements:

* ConsumerId MUST be of type String.
* ConsumerId MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* ConsumerId MUST adhere to the following nullability requirements:
  * ConsumerId MUST be null when a [*charge*](#glossary:charge) is not attributed to a specific *consumer* by the service provider.
  * ConsumerId MUST NOT be null when a *charge* is attributed to a specific *consumer* by the service provider.
* When ConsumerId is not null, ConsumerId MUST adhere to the following requirements:
  * ConsumerId MUST be a unique identifier within the context of the service provider.

>   * ConsumerId MUST NOT contain plain-text personally identifiable information (PII) if the service provider supplies an alternative opaque identifier.

Reviewer: ```suggestion
* ConsumerId MUST adhere to the following nullability requirements:
  * ConsumerId MUST be null when the [*service provider*](#glossary:service-provider) cannot associate a [*charge*](#glossary:charge) with a *consumer*.
  * ConsumerId MUST NOT be null when the *service provider* can associate a *charge* with a *consumer*, regardless of whether the *consumer* was provider-identified or customer-supplied.
* When ConsumerId is not null, ConsumerId MUST adhere to the following requirements:
  * ConsumerId MUST be a unique identifier within the *service provider* context.
  * ConsumerId MUST NOT contain plain-text personally identifiable information (PII) when an opaque *consumer* identifier is available.
```

Author: Going to hold on making any further changes to nullability until there are wider TF and Maintainer discussions about this.

Reviewer 2: The `ConsumerId` work has moved out of this PR. Per the TF2 and Maintainers decision to deliver the actor columns incrementally, #2360 is now scoped to `PrincipalId` only, and `ConsumerId` (including this nullability question) has been carved out to draft PR #2495, stacked on this branch.

Continuing this discussion there, re-anchored on the `consumerid.md` nullability requirements: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2495#discussion_r3496277075

## Column ID

ConsumerId

## Display Name

Consumer ID

## Description

> Unique identifier of the downstream entity or end-actor consuming the resource or service.

Reviewer 2: ```suggestion
Unique identifier of the downstream entity or end-actor consuming the *resource* or *service*.
```

Italicizing glossary terms in the Description section is consistent with the convention in neighboring columns (e.g., [resourceid.md](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/resourceid.md), [consumedunit.md](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/consumedunit.md)). Glossary links already appear on first occurrence in the opening paragraph, so the link-once rule is satisfied with italics-only here.

> Unique identifier of the downstream entity or end-actor consuming the *resource* or *service*.

Reviewer: ```suggestion
Identifier representing the downstream actor that consumed a *resource* or *service*.
```

# Consumer ID

A Consumer ID is an identifier representing the [*consumer*](#glossary:consumer): a discrete downstream actor (e.g., end-user, application, autonomous agent) that consumed a [*resource*](#glossary:resource) or [*service*](#glossary:service). The Consumer ID is commonly used for cost allocation, showback/chargeback, and unit economics calculations without exposing raw actor data.

> Consumer ID is used in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Consumer ID represents the downstream actor that consumed the resource or service, Principal ID represents and authorized actor that initiated a request.

Author: ```suggestion
Consumer ID is used in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Consumer ID represents the downstream actor that consumed the resource or service, Principal ID represents the authorized actor that initiated a request.
```

> Consumer ID is used in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Consumer ID represents the downstream actor that consumed the resource or service, Principal ID represents the authenticated and authorized actor that initiated a request.

Reviewer: Per the May 19 ad-hoc discussion, "authorized" will be used for principals, implicitly including authentication. This change was already applied to the glossary and principal introduction section; this update aligns with that terminology.

```suggestion
Consumer ID is used in conjunction with [Principal ID](#datasets.costandusage.principalid) to resolve asymmetric actor granularity. While Consumer ID represents the downstream actor that consumed the resource or service, Principal ID represents and authorized actor that initiated a request.
```
